### DISCUSSION TITLE
About Road Map

### DISCUSSION THREAD

**Discussion by ZLY201** [ZLY201/typescript-tutorial-exercises#35] [Ideas]
## Feature 1

The puzzle set is the core of the project. We need more high-quality and entry-friendly questions to help developers learn.

### Tutorial exercises
Help developers start from scratch. Different topics correspond to different chapters here.

### Puzzle
After completing the tutorial, developers can consider challenging more difficult questions

The good news is that for Tutorial, we have @mdevils to help, and [his project](https://github.com/typescript-exercises/typescript-exercises) has accumulated dozens of high-quality exercises.

For puzzles, we can choose from [types challenge](https://github.com/type-challenges/type-challenges).

## Feature 2

The puzzle set should not only include TS-exercise but also help to learn how to use tsconfig. So we should add a function like [playground](https://www.typescriptlang.org/play).

![image](https://github.com/typescript-tutorial-exercises/core/assets/59038614/95d58fbf-e7e6-478d-aaf5-911695d77133)

But we can only add a json editor to monaco.

![image](https://github.com/typescript-tutorial-exercises/core/assets/59038614/7b4cfacf-50d4-4cbd-82b1-7345bd160f89)

## Feature 3

Open discussion forums on your page to make it feel more like a community.

Type-challenges use the issues of github to implement this function, I think we should implement at the exercise page.

![image](https://github.com/typescript-tutorial-exercises/core/assets/59038614/016756d3-2e1d-466b-8e6a-65297d949fa9)

**Comment by ZLY201** [comment]
To digress, I don’t know much about the aesthetic requirements of other countries. I have implemented all the pages according to my inherent ideas. Do I need to make any corrections?

**Comment by ZLY201** [comment]
I can help to implement Feature 2 and 3 but I'm not very good at asking questions.

**Comment by mdevils** [comment]
My thoughts:

1. Good job so far!

2. Teaching people to use tsconfig is important, but if we compare the important of tsconfig with just TS syntax/typing, then TS itself should be more important. Therefore I suggest to first focus on TS itself before going further and playing with tsconfig.

3. About puzzles.
    > For puzzles, we can choose from [types challenge](https://github.com/type-challenges/type-challenges).

    Not sure if we can copy-paste from those challenges, but we can use them for inspiration.

4. I suggest not to store solutions in the app itself, but instead use `github issues` like [types challenges](https://github.com/type-challenges/type-challenges) project does. This allows users to post and share their solutions with community. And there can be different solutions to one problem. Example: https://github.com/type-challenges/type-challenges/issues?q=label%3A14188+label%3Aanswer+sort%3Areactions-%2B1-desc In order to implement it we need two buttons:
    a) `[Find solutions]` which leads to `github issues` pages with a filter for the specified challenge/puzzle.
    b) `[Share your solution]` which takes user's code and opens github new issue page with labels, title and description prefilled with user's code.

5. Remove `Records` page. Because it's not really a good use-case for TS exercises, we don't compete in performance. And shorter doesn't mean better. We should rather focus on allowing user to see other users' solutions (see `[Find solutions]` button proposal).

6. Put difficulties to the challenges. I.e. `easy`, `medium`, `hard` and make them visible to the user.

7. Don't allow user to delete test cases. There is just no need for that in case of TS.

8. Check solution automatically. I don't think we need `[Run]` button in case of TS types challenges.

9. There is a big problem with run errors. Every time I want to check my code, I get very unclear error messages. Take a look at `typescript-exercises`, there you see the test file and when you get an error, it sends to to the specific line/test in that test file or the exercise file.

10. Regarding "Feature 3: discussions". I'd recommend to stay server-less and store as many things as possible in github.

**Reply by ZLY201** [reply]
I agree most of yours but three things should be talked:
 
1. Github issues can be fetched by using some plugin like Gitalk. So we can open the `discussions` page by using Github issues.
2. The cases is necessary because for some puzzle like `type Transform<T> = Other<T>` we need let user try there own cases to debug. E.g. https://typescript-tutorial-exercises.github.io/?problem=TransposeMatrix
3. About the run errors, I don't have a good idea to optimize. The only way I can think of is this is to show the `check.ts` file on the page.

**Comment by ZLY201** [comment]
I refactored all the code, now it is just a visual version of type-challenges, in addition to the known bugs, we still have some things to move forward:
1. Editor multi-file viewing
2. Compile errors can jump
3. Solution to solutions
